### PR TITLE
Add footnote support for wiki markdown

### LINKED ISSUE #7316: Superscript support for osu! wiki (footnotes)
the title says it all, but in case it does not: we'd *love* support of superscript formatting (ᵗʰᵃᵗ ᶠˡᵒᵃᵗᶦⁿᵍ ᵗᵉˣᵗ ᵗʰᶦⁿᵍ ᵖᵉᵒᵖˡᵉ ᵘˢᵉ ᶠᵒʳ ᶠᵒᵒᵗⁿᵒᵗᵉˢ) in Markdown for osu! wiki to have it look like Wikipedia ([quick example: search for `[1]` here](https://en.wikipedia.org/wiki/Map_symbol#Representing_spatial_phenomena)).

could be anything really, but I've seen this in one of Markdown parsers:

```markdown
blah blah^[1] blah^[anything, not only numbers] ...

...

[1]: http://bl.ah?
[anything, not only numbers]: https://bl.ah/blah
```

### PR DESCRIPTION
**Pull request by gagahpangeran**
- Closes #7316 
- [ ] Depends on #7980

## How to use

Basic syntax:

```markdown
Footnote[^ref]

[^ref]: Reference here
```

And will automatically assign number and render something like this:

Footnote<sup><a href="#">[1]</a></sup>

1. Reference here <a href="#">↑</a>

[1] link will target to the reference list item and ↑ link will go back to [1].

### Some caveats

**Be careful of exclamation**

Valid:

```markdown
osu! [^ref]

[^ref]: https://osu.ppy.sh
```

Invalid:

```markdown
osu![^ref]

[^ref]: https://osu.ppy.sh
```

**Be careful of duplicate reference key**

Valid:

```markdown
First[^ref]

Second[^ref]

[^ref]: Valid shared reference
```

Invalid:

```markdown
First[^ref]

Second[^ref]

[^ref]: This become invalid
[^ref]: Valid shared reference
```

## Demo

*[This file](https://raw.githubusercontent.com/gagahpangeran/osu-wiki/test-footnote/wiki/Welcome/en.md) is used for this demo*

https://user-images.githubusercontent.com/4964985/128446811-68b89d64-6305-4e38-8c30-d1f1646380b7.mp4

### CODE REVIEW

File: app/Libraries/Markdown/Indexing/Extension.php [hunk 1]
```diff
@@ -9,6 +9,7 @@
 use League\CommonMark\Block\Element as Block;
 use League\CommonMark\ConfigurableEnvironmentInterface;
 use League\CommonMark\Extension\ExtensionInterface;
+use League\CommonMark\Extension\Footnote\Node as FootnoteExtension;
```
Comment (nanaya) [inline app/Libraries/Markdown/Indexing/Extension.php:12]: seeing how many renderers are being overridden here (are they even necessary?) please look into updating commonmark to latest version (2.0) first.

**Review by nanaya — CHANGES_REQUESTED** [review]

File: app/Libraries/Markdown/Indexing/Extension.php (hunk 1 above)
Comment (gagahpangeran) [reply app/Libraries/Markdown/Indexing/Extension.php:12]: - `FootnoteContainerRenderer` is used to match `osu-md__list` structure.
- `FootnoteListItemRenderer` is used to match `osu-md__list-item` structure.
- `FootnoteRefRenderer` and `FootnoteRefRenderer` is used to tweak the content a bit.

**Comment by gagahpangeran** [comment]
Looks like updating commonmark to version 2.0 will took a while and will be in the separated PR.

I already start updating it and pretty much breaking changes everywhere (So much API and namespace changes). This looks fun 🙃

**Comment by nanaya** [comment]
@gagahpangeran, are you still working on updating commonmark?

**Comment by gagahpangeran** [comment]
> @gagahpangeran, are you still working on updating commonmark?

Yeah, the progress is about 90%.

Maybe will open PR tomorrow.

**Comment by gagahpangeran** [comment]
Closing form now.

Will create new PR after #8087 is merged.

**Pull request closed without merging**